include/ssd-internal.h

```diff
@@ -0,0 +1,139 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+#ifndef __LABWC_SSD_INTERNAL_H
```

Comment:
For `src/action.c` I used `include/private/action.h`, but I actually tend to like the `-internal` variant better (e.g. `include/action-internal.h`).

We should likely discuss having the same place and name + `-internal.h` or having it sub-directory based (e.g. `include/internal/ssd.h`), write that down in Contributing.md and then migrate both cases to whatever we decide on. Pinging @johanmalm as well.

Reply:
I like to avoid having two headers with the same name -- it tends to confuse IDEs. I can go ahead and rename `include/private/action.h` to `include/action-internal.h` if you're good with that.

Reply:
Looking at it again, `include/private/action.h` was included only in `action.c`, so I just inlined it there,

Reply:
Looks like there are no more `include/private/` then. I thought we had a 'private' header file common to the files in `src/ssd`, but can't find it now. Maybe my memory is playing tricks on me.

I don't feel strongly on this, but my preference is as follows in terms of location/naming of 'private' header files (starting with highest preference):

1. Put 'private' header files in `src/<subdir-that-wants-common-header>` using the format `foo_p.h`. This seems clean and self-contained, but we lose the ability to grep `include/` for function declarations (which isn't a big deal for me).
2. Put 'private' header files in 'include/' and calling them `foo_p.h`
3. Same as (3) but call them `foo-internal.h`

Reply:
I am fine with any of these, though (1) is a little bit unclear when there isn't a subdirectory; for example I am considering adding a `view-internal.h` that would be used from `view.c`, `view-impl.c`, `xdg.c`, and `xwayland.c`.

Reply:
@Consolatis and I pondered these private files before and not really a clean approach that fits all.

Sway often has larger .c files and thus less need for private header files.

At the end of the day, it’s not a library so we don’t have to be super strict about public/private. So it comes down to readability/maintainability.


The xbm/ dir is a good example. We should only really ‘expose’ one function in the include/ header. The rest is local implementation detail.
Sway probably would have put it all in one .c file. 

I think that when it’s starts getting a bit grey it’s probably a good time to stop splitting files 😄 

Whats the common view/xdg/xwayland implementation detail? Are you thinking of the structs?

Reply:
> Whats the common view/xdg/xwayland implementation detail? Are you thinking of the structs?

What I am hoping to do is to make `struct view` effectively read-only outside `view.c` and friends, so as to enforce that e.g. `view->x` can't be overwritten from `desktop.c` directly; it has to call `view_move()` or similar.

So I was imagining something along the lines of:
- Make `struct view` itself opaque (outside `view-internal.h`).
- Put most of the flags/geometry/etc. in a nested struct (`view_state` or such).
- Then have an accessor `const view_state *view_get_state(view *)`. This would allow broad read-only access to e.g. `view->state->x` or `view->state->maximized` without allowing those fields to be overwritten directly.

I haven't fully fleshed out the idea, and it may or may not still make sense once #641 lands.

Reply:
> Put 'private' header files in src/<subdir-that-wants-common-header> using the format foo_p.h. This seems clean and self-contained, but we lose the ability to grep include/ for function declarations (which isn't a big deal for me).

It would be for my workflow, I'd really prefer all the headers being in `include/`.

> for example I am considering adding an include/view-internal.h that would be used from view.c, view-impl.c, xdg.c, and xwayland.c.

Which raises the question if we want a "module" directory for those (+ `xwayland-unmanaged.c`), similar to `src/ssd/`. I may be overshooting with this though.

Reply:
A `view` subdirectory might be nice in future.

For this MR, is it okay to stick with `include/ssd-internal.h` as it stands currently?

Reply:
Yes, fine.